doc/drop_counters/drop_counters_HLD.md

```diff
@@ -0,0 +1,353 @@
+* [6 Unit Tests](#6-unit-tests)
+* [7 Open Questions](#7-open-questions)
+
+
+
+# List of Tables
+* [Table 1: Abbreviations](#definitionsabbreviation)
+
+# List of Figures
+* [Figure 1: General Flow](#41-general-flow)
+
+# Revision
+| Rev | Date     | Author      | Change Description |
+|:---:|:--------:|:-----------:|--------------------|
+| 0.1 | 07/30/19 | Danny Allen | Initial version    |
+| 0.2 | 09/03/19 | Danny Allen | Review updates     |
+
+# About this Manual
+This document provides an overview of the implementation of configurable packet drop counters in SONiC.
+
+# Scope
+This document describes the high level design of the configurable drop counter feature.
+
+# Definitions/Abbreviation
+| Abbreviation | Description     |
+|--------------|-----------------|
+| RX           | Receive/ingress |
+| TX           | Transmit/egress |
+
+# 1 Overview
+The main goal of this feature is to provide better packet drop visibility in SONiC by providing a mechanism to count and classify packet drops that occur due to different reasons. 
+
+The other goal of this feature is for users to be able to track the types of drop reasons that are important for their scenario. Because different users have different priorities, and because priorities change over time, it is important for this feature to be easily configurable.
+
+We will accomplish both goals by adding support for SAI debug counters to SONiC. 
+* Support for creating and configuring port-level and switch-level debug counters will be added to orchagent and syncd. 
+* A CLI tool will be provided for users to manage and configure their own drop counters
+
+# 2 Requirements
+
+## 2.1 Functional Requirements
+1. CONFIG_DB can be configured to create debug counters
+2. STATE_DB can be queried for debug counter capabilities
+3. Users can access drop counter information via a CLI tool
+    1. Users can see what capabilities are available to them
+        1. Types of counters (i.e. port-level and/or switch-level)
+        2. Number of counters
+        3. Supported drop reasons
+    2. Users can see what types of drops each configured counter contains
+    3. Users can add and remove drop reasons from each counter
+    4. Users can read the current value of each counter
+    5. Users can assign aliases to counters
+    6. Users can clear counters
+
+## 2.2 Configuration and Management Requirements
+Configuration of the drop counters can be done via:
+* config_db.json 
+* minigraph.xml
+* CLI
+
+## 2.3 Scalability Requirements
+Users must be able to use all counters and drop reasons provided by the underlying hardware.
+
+## 2.4 Supported Debug Counters
+* PORT_INGRESS_DROPS: port-level ingress drop counters
+* PORT_EGRESS_DROPS: port-level egress drop counters
+* SWITCH_INGRESS_DROPS: switch-level ingress drop counters
+* SWITCH_EGRESS_DROPS: switch-level egress drop counters
+
+# 3 Design
+
+## 3.1 CLI (and usage example)
+The CLI tool will provide the following functionality:
```

Comment:
is there any default configurations for drop counters? 
do you have any template to confiugre the debug counters?

Reply:
We decided during the meeting yesterday that this is out of the scope of the current design. This is a good idea though and I think we'll address it during the discussions about the management framework or potentially as a future extension to this feature.